File: src/dataset/python/allergies_exercism-ThomasZumsteg/gemini/GeminiFlash_allergies_v3.py
"""Tracks allergies"""

class Allergies(object):
    """Generates and stores allergies from a given score"""

    allergen_map = {
        1: 'eggs',
        2: 'peanuts',
        4: 'shellfish',
        8: 'strawberries',
        16: 'tomatoes',
        32: 'chocolate',
        64: 'pollen',
        128: 'cats',
    }

    def __init__(self, score):
        """Generates a list of allergies from a score"""
        self.lst = []
        remaining_score = score
        for value in sorted(Allergies.allergen_map.keys(), reverse=True):
            if remaining_score >= value:
                self.lst.append(Allergies.allergen_map[value])
                remaining_score -= value
        self.lst.reverse()


    def is_allergic_to(self, allergie):
        """Tests if allergie is on the list"""
        return allergie in self.lst

File: src/dataset/cpp/allergies_exercism-S0l0m4n/gemini/GeminiFlash_allergies_v3.cpp
/**
 * Given a person's allergy score, determine whether or not they're allergic to
 * a specific allergy and their full list of allergies.
 */

#include "allergies.h"

#include <algorithm>
#include <array>
#include <numeric>

using std::string;
using std::unordered_set;

namespace allergies {

    constexpr std::array<const char*, 8> allergy_names = {
        "eggs",
        "peanuts",
        "shellfish",
        "strawberries",
        "tomatoes",
        "chocolate",
        "pollen",
        "cats"
    };

    allergy_test::allergy_test(int score_value) : score(score_value) {
        int remaining_score = score_value;
        for (size_t i = 0; i < allergy_names.size(); ++i) {
            int allergy_score = 1 << i;
            if (remaining_score >= allergy_score) {
                if (remaining_score & allergy_score) {
                    allergy_set.insert(allergy_names[i]);
                    remaining_score -= allergy_score;
                }
            } else {
                break; // Optimization: No need to check further if allergy_score exceeds remaining_score
            }
        }
    }

    bool allergy_test::is_allergic_to(string allergy_name) {
        return allergy_set.count(allergy_name) > 0;
    }

    unordered_set<string> allergy_test::get_allergies() {
        return allergy_set;
    }
}  // namespace allergies